File: KantaroWpf/Converters/StrofeFlowDocumentConverter.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Windows.Data;
using System.Windows.Documents;
using System.Windows;

namespace Enzo.Music.KantaroWpf.Converters;

public partial class StrofeFlowDocumentConverter : IValueConverter
{

    private const string LinkPrefix = "kanto://accordo#";

    private string LinkFormat
    {
        get
        {
            var sb = new StringBuilder();
            sb.Append(LinkPrefix);
            sb.Append("{0}");
            return sb.ToString();
        }
    }

    public static string? GetAccordoFromUri(Uri navigateUri)
    {
        if (navigateUri is null)
            return null;
        // Verifica la corretta valorizzazione dell'uri passato come parametro
        if (!navigateUri.AbsolutePath.StartsWith(LinkPrefix))
        {
            throw new InvalidCastException("Unrecognized uri format");
        }
        return navigateUri.AbsolutePath.Remove(0, LinkPrefix.Length);
    }

    public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
    {
        if (value is null || !(value is IList<Strofa>))
            return DependencyProperty.UnsetValue;
        if (targetType != typeof(FlowDocument))
        {
            throw new InvalidOperationException("Il tipo di destinazione deve essere un FlowDocument");
        }
        var doc = new FlowDocument();
        IList<Strofa> strofe = (IList<Strofa>)value;
        foreach (var strofa in strofe)
        {
            var b = new Paragraph();
            if (strofa.Nome is not null)
            {
                b.Tag = strofa.Nome;
                b.FontStyle = FontStyles.Italic;
            }
            if (strofa is StrofaRipetuta)
            {
                b.Padding = new Thickness(10, 5, 0, 5);
                b.FontWeight = FontWeights.Bold;
                b.FontStyle = FontStyles.Italic;
                var lnk = new Hyperlink(new Run(((StrofaRipetuta)strofa).Riferimento));
                b.Inlines.Add(lnk);
            }
            else
            {
                foreach (var p in strofa.Parti)
                {
                    Inline tb;
                    if (p is not null)
                    {
                        if (p.Accordo is null)
                        {
                            tb = new Run(p.Testo);
                        }
                        else
                        {
                            tb = new Hyperlink(new Run(p.Testo));
                            ((Hyperlink)tb).NavigateUri = new Uri(string.Format(LinkFormat, p.Accordo.ToString()));
                        }
                    }
                    else
                    {
                        tb = new LineBreak();
                    }
                    b.Inlines.Add(tb);
                }
            }
            doc.Blocks.Add(b);
        }
        return doc;
    }

    public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
    {
        return DependencyProperty.UnsetValue;
    }
}

File: CanzoniereWPF/Converters/StrofeFlowDocumentConverter.vb
Public Class StrofeFlowDocumentConverter
    Implements IValueConverter

    Private Const LinkPrefix As String = "kanto://accordo#"

    Private ReadOnly Property LinkFormat As String
        Get
            Dim sb As New Text.StringBuilder
            sb.Append(StrofeFlowDocumentConverter.LinkPrefix)
            sb.Append("{0}")
            Return sb.ToString
        End Get
    End Property

    Public Shared Function GetAccordoFromUri(navigateUri As Uri) As String
        If navigateUri Is Nothing Then Return Nothing
        ' Verifica la corretta valorizzazione dell'uri passato come parametro
        If Not navigateUri.AbsolutePath.StartsWith(StrofeFlowDocumentConverter.LinkPrefix) Then
            Throw New InvalidCastException("Unrecognized uri format")
        End If
        Return navigateUri.AbsolutePath.Remove(0, LinkPrefix.Length)
    End Function

    Public Function Convert(value As Object, targetType As Type, parameter As Object, culture As Globalization.CultureInfo) As Object Implements IValueConverter.Convert
        If value Is Nothing OrElse Not (TypeOf value Is IList(Of Strofa)) Then Return DependencyProperty.UnsetValue
        If targetType <> GetType(FlowDocument) Then
            Throw New InvalidOperationException("Il tipo di destinazione deve essere un FlowDocument")
        End If
        Dim doc As New FlowDocument
        Dim strofe = DirectCast(value, IList(Of Strofa))
        For Each strofa In strofe
            Dim b As New Paragraph()
            If strofa.Nome IsNot Nothing Then
                b.Tag = strofa.Nome
                b.FontStyle = Windows.FontStyles.Italic
            End If
            If TypeOf strofa Is StrofaRipetuta Then
                b.Padding = New Thickness(10, 5, 0, 5)
                b.FontWeight = FontWeights.Bold
                b.FontStyle = FontStyles.Italic
                Dim lnk As New Hyperlink(New Run(DirectCast(strofa, StrofaRipetuta).Riferimento))
                b.Inlines.Add(lnk)
            Else
                For Each p In strofa.Parti
                    Dim tb As TextElement
                    If p IsNot Nothing Then
                        If p.Accordo Is Nothing Then
                            tb = New Run(p.Testo)
                        Else
                            tb = New Hyperlink(New Run(p.Testo))
                            DirectCast(tb, Hyperlink).NavigateUri = New Uri(String.Format(LinkFormat, p.Accordo.ToString))
                        End If
                    Else
                        tb = New LineBreak
                    End If
                    b.Inlines.Add(tb)
                Next
            End If
            doc.Blocks.Add(b)
        Next
        Return doc
    End Function

    Public Function ConvertBack(value As Object, targetType As Type, parameter As Object, culture As Globalization.CultureInfo) As Object Implements IValueConverter.ConvertBack
        Return DependencyProperty.UnsetValue
    End Function
End Class
